src/lib/listeners/ProgressEventListener.ts

```diff
@@ -0,0 +1,192 @@
+		this.display = display;
+		this.tickTime = tickTime;
+		this._isSpinning = false;
+	}
+
+	protected startSpinning(action: string): void {
+		/* istanbul ignore if: Shouldn't be possible for spinner to start twice */
+		if (this._isSpinning) {
+			return;
+		}
+		this.startTime = Date.now();
+		this.display.spinnerStart(action, this.createSpinnerStatus());
+		if (this.tickTime > 0) {
+			// `setInterval` means the callback will be called repeatedly. This allows us to automatically refresh the
+			// spinner on a regular interval even if nothing happened. This is primarily useful for incrementing a timer,
+			// so the user doesn't feel like the system is frozen.
+			this.tickIntervalId = setInterval(() => {
+				this.display.spinnerUpdate(this.createSpinnerStatus());
+			}, this.tickTime);
+		}
+		this._isSpinning = true;
+	}
+
+	protected updateSpinner(): void {
+		this.display.spinnerUpdate(this.createSpinnerStatus());
+	}
+
+	protected stopSpinning(): void {
+		if (this.tickIntervalId) {
+			clearInterval(this.tickIntervalId);
+		}
+		if (this._isSpinning) {
+			this.display.spinnerStop(getMessage(BundleName.ProgressEventListener, 'base-spinner.done'));
+			this._isSpinning = false;
+		}
+	}
+
+	protected isSpinning(): boolean {
+		return this._isSpinning;
+	}
+
+	/**
+	 * Return the number of seconds (rounded down) that the spinner has been spinning.
+	 * @protected
+	 */
+	protected getSecondsSpentSpinning(): number {
+		return Math.floor((Date.now() - this.startTime) / 1000);
+	}
+
+	protected abstract createSpinnerStatus(): string;
+}
+
+export class RuleSelectionProgressSpinner extends ProgressSpinner implements ProgressEventListener {
+	private isListening: boolean;
+	private engineNames: string[];
+	private completionPercent: number;
+
+	/**
+	 * @param display The display with which to show output to the user
+	 * @param tickTime A number of milliseconds to wait between automatic ticks. Defaults to 1000. Negative values disable ticking.
+	 */
+	public constructor(display: Display, tickTime: number = 1000) {
+		super(display, tickTime);
+		this.isListening = false;
+	}
+
+	public listen(codeAnalyzer: CodeAnalyzer): void {
+		/* istanbul ignore else: should not be calling listen() twice */
+		if (!this.isListening) {
+			this.isListening = true;
+			this.engineNames = codeAnalyzer.getEngineNames();
+			this.completionPercent = 0;
+			codeAnalyzer.onEvent(EventType.RuleSelectionProgressEvent, (e: RuleSelectionProgressEvent) => this.handleEvent(e));
+		}
+	}
+
+	public stopListening(): void {
+		if (this.isListening) {
+			this.isListening = false;
+			this.stopSpinning();
+		}
+	}
+
+	private handleEvent(e: RuleSelectionProgressEvent): void {
+		if (!this.isListening) {
+			return;
+		}
+
+		if (!this.isSpinning()) {
+			this.startSpinning(getMessage(BundleName.ProgressEventListener, 'selection-spinner.action'));
+		}
+
+		this.completionPercent = Math.floor(e.percentComplete);
```

Comment:
I'm using `Math.floor` here, because it's visually unpleasant to have the percentage go back and forth between whole numbers and long decimals (e.g., from `20%` to `40.666666666666667%` to `60%`).
If possible, it might be optimal for Core to apply `Math.floor` and guarantee that the percentages it sends out are whole numbers.